### ISSUE TITLE
Ajax: success callback triggered instead of error when call fails because user offline

### ISSUE THREAD

**Issue by mjrusso** [madrobby/zepto#283]
When a network request is made and there is no connectivity, the Ajax success callback is triggered. However, this is an error scenario and should be treated as such (i.e., the error callback should be triggered).

Although this problem has existed for quite some time, it was less likely to be noticed with older versions of Zepto.

An older version of code looked like this:

``` javascript
xhr.onreadystatechange = function(){
   if (xhr.readyState == 4) {
     var result, error = false;
     if ((xhr.status >= 200 && xhr.status < 300) || xhr.status == 0) {
       if (mime == 'application/json') {
         try { result = JSON.parse(xhr.responseText); }
         catch (e) { error = e; }
       }
       else result = xhr.responseText;
       if (error) settings.error(xhr, 'parsererror', error);
       else settings.success(result, 'success', xhr);
     } else {
       error = true;
       settings.error(xhr, 'error');
     }
     settings.complete(xhr, error ? 'error' : 'success');
   }
};
```

With this code, if a user is offline and makes an network request for a JSON resource, the error handler would be triggered. (The error would be reported as a `parseerror`, which is not really correct, but at least the error handler is invoked.) More specifically, `xhr.status` would be 0 in this case, and the JSON parser would attempt to parse an empty string, which would throw an exception.

However, the latest Ajax implementation now looks like this:

``` javascript
xhr.onreadystatechange = function(){
  if (xhr.readyState == 4) {
    clearTimeout(abortTimeout);
    var result, error = false;
    if ((xhr.status >= 200 && xhr.status < 300) || xhr.status == 0) {
      if (mime == 'application/json' && !(/^\s*$/.test(xhr.responseText))) {
        try { result = JSON.parse(xhr.responseText); }
        catch (e) { error = e; }
      }
      else result = xhr.responseText;
      if (error) settings.error.call(context, xhr, 'parsererror', error);
      else settings.success.call(context, result, 'success', xhr);
    } else {
      error = true;
      settings.error.call(context, xhr, 'error');
    }
    settings.complete.call(context, xhr, error ? 'error' : 'success');
  }
};
```

With the addition of `!(/^\s*$/.test(xhr.responseText)`, empty responses are no longer parsed as JSON. **However, the request is assumed to be successful.**

Digging through the Zepto source, I found the commit (>1 year old) that started treating the status of 0 as a success (SHA: ce9f1f3ac5d42dc68a03c9a940ffbba9437295ca; _"Make Ajax calls work for local files (file:///...)."_

jQuery ([Ajax source](https://github.com/jquery/jquery/blob/master/src/ajax.js)) treats a status of 0 as an error in all conditions, unless the request is for a local resource.

Here's the relevant code for remote requests:

``` javascript
if ( status >= 200 && status < 300 || status === 304 ) {
  // ...
} else {
  // this is treated as an error
}
```

And the local-specific code:

``` javascript
// If the request is local and we have data: assume a success
// (success with no data won't get notified, that's the best we
// can do given current implementations)
if ( !status && s.isLocal && !s.crossDomain ) {
    status = responses.text ? 200 : 404;
}
```

So, in the jQuery case, if a request is made for a remote resource and the user is offline, the error callback is properly triggered. The only scenario that is not properly handled is if a request is made for a local file and the file exists but has a length of 0 -- this is treated like a 404.

To summarize, I think that it is very important to trigger an error in all scenarios if `xhr.status` is 0 and the request is made for a remote resource. For requests to local files, the behavior could remain the same (treat all local requests as successful, even responses with empty bodies, given that it is not possible to distinguish between a local file that exists but is empty, and a file that does not exist).

**Comment by joantune** [comment]
Hi fellas!

I bumped again against this issue.

TL;DR: The ```protocol``` variable should refer to the URL of the XHR request, not the current URL.

The why is:

In the scenario where you are using Zepto on a local file to make an Ajax request to a remote webserver (which might sound weird, but it is quite common with Cordova/Phonegap apps), if we have no connectivity the status is 0, but the protocol is file, thus we don't get our error callback as we should.

**Comment by mislav** [comment]
@joantune Thanks for the heads up. Can you please submit a PR to work around this problem?

**Comment by joantune** [comment]
@mislav (and everyone that comes here) sorry, please disregard my post. As I was looking now a bit more to it, to try to come up with a PR, I realized that the protocol variable did what I stated, plus, I debugged it further and the error callback was called with the original version! I'm sorry about the noise that I introduced. 
Today it is a bit late around here, I'll look it up a bit better tomorrow, If there's actually something wrong with the framework, I'll introduce a PR. Sorry again.

**Comment by mislav** [comment]
No probs; thanks for the update.
